File: src/Sample/Plumix.Sample/Demos/General/NavigationPopDemoPage.cs
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/navigation_pop_demo_page.dart (exact sample parity)

public sealed class NavigationPopDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new NavigationPopDemoPageState();
    }
}

internal sealed class NavigationPopDemoPageState : State
{
    private NavigatorState? _nestedNavigator;
    private bool _canLeave = true;
    private int _nestedPage = 1;
    private string _status = "No pop attempted";

    public override Widget Build(BuildContext context)
    {
        return new PopScope<object?>(
            canPop: _canLeave,
            onPopInvokedWithResult: (didPop, result) =>
            {
                if (Mounted)
                {
                    SetState(() => _status = didPop ? "Route popped" : "Pop handled or blocked");
                }
            },
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 12,
                children:
                [
                    new Text("PopScope + NavigatorPopHandler", fontSize: 20, color: Colors.Black),
                    new Text(
                        "Push a nested page, then simulate a parent Back. The handler consumes it in the nested " +
                        "navigator. Disable route pop to probe PopScope veto behavior.",
                        fontSize: 14,
                        color: Colors.DimGray),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            new TextButton(
                                onPressed: ToggleCanLeave,
                                child: new Text(_canLeave ? "Disable route pop" : "Enable route pop")),
                            new TextButton(
                                onPressed: PushNestedPage,
                                child: new Text("Push nested page")),
                            new TextButton(
                                onPressed: () => Navigator.Of(Context).MaybePop("demo-result"),
                                child: new Text("Simulate parent Back")),
                        ]),
                    new Text($"Status: {_status}", color: Color.Parse("#FF31506F")),
                    new Expanded(
                        child: new NavigatorPopHandler<object?>(
                            onPopWithResult: result =>
                            {
                                _nestedNavigator?.MaybePop(result);
                                if (Mounted)
                                {
                                    SetState(() => _status = "NavigatorPopHandler popped the nested route");
                                }
                            },
                            child: new Navigator(
                                initialRoute: new BuilderPageRoute(
                                    builder: nestedContext =>
                                    {
                                        _nestedNavigator = Navigator.Of(nestedContext);
                                        return BuildNestedPage(0);
                                    },
                                    settings: new RouteSettings(Name: "nested-root"))))),
                ]));
    }

    private void ToggleCanLeave()
    {
        SetState(() =>
        {
            _canLeave = !_canLeave;
            _status = _canLeave ? "Route pop enabled" : "Route pop disabled";
        });
    }

    private void PushNestedPage()
    {
        int page = _nestedPage++;
        _nestedNavigator?.Push(new BuilderPageRoute(
            builder: _ => BuildNestedPage(page),
            settings: new RouteSettings(Name: $"nested-{page}")));
        SetState(() => _status = $"Nested page {page} pushed");
    }

    private static Widget BuildNestedPage(int page)
    {
        return new Container(
            color: page == 0 ? Color.Parse("#FFE8F0FE") : Color.Parse("#FFE6F4EA"),
            padding: new Thickness(16),
            alignment: Alignment.Center,
            child: new Text(
                page == 0 ? "Nested root" : $"Nested page {page}",
                fontSize: 18,
                color: Colors.Black));
    }
}

File: dart_sample/lib/demos/general/navigation_pop_demo_page.dart
import 'package:flutter/material.dart';

class NavigationPopDemoPage extends StatefulWidget {
  const NavigationPopDemoPage({super.key});

  @override
  State<NavigationPopDemoPage> createState() => _NavigationPopDemoPageState();
}

class _NavigationPopDemoPageState extends State<NavigationPopDemoPage> {
  final GlobalKey<NavigatorState> _nestedNavigatorKey =
      GlobalKey<NavigatorState>();
  bool _canLeave = true;
  int _nestedPage = 1;
  String _status = 'No pop attempted';

  @override
  Widget build(BuildContext context) {
    return PopScope<Object?>(
      canPop: _canLeave,
      onPopInvokedWithResult: (bool didPop, Object? result) {
        if (mounted) {
          setState(() {
            _status = didPop ? 'Route popped' : 'Pop handled or blocked';
          });
        }
      },
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 12,
        children: <Widget>[
          const Text(
            'PopScope + NavigatorPopHandler',
            style: TextStyle(fontSize: 20, color: Colors.black),
          ),
          const Text(
            'Push a nested page, then simulate a parent Back. The handler consumes it in the nested '
            'navigator. Nested routes use PageTransitionsTheme, including Android predictive-back '
            'peek/commit/cancel and iOS/macOS leading-edge swipe. Disable route pop to probe PopScope '
            'veto behavior.',
            style: TextStyle(fontSize: 14, color: Colors.black54),
          ),
          Row(
            spacing: 8,
            children: <Widget>[
              TextButton(
                onPressed: _toggleCanLeave,
                child: Text(
                  _canLeave ? 'Disable route pop' : 'Enable route pop',
                ),
              ),
              TextButton(
                onPressed: _pushNestedPage,
                child: const Text('Push nested page'),
              ),
              TextButton(
                onPressed: () =>
                    Navigator.of(context).maybePop<Object?>('demo-result'),
                child: const Text('Simulate parent Back'),
              ),
            ],
          ),
          Text(
            'Status: $_status',
            style: const TextStyle(color: Color(0xFF31506F)),
          ),
          Expanded(
            child: NavigatorPopHandler<Object?>(
              onPopWithResult: (Object? result) {
                _nestedNavigatorKey.currentState?.maybePop<Object?>(result);
                if (mounted) {
                  setState(() {
                    _status = 'NavigatorPopHandler popped the nested route';
                  });
                }
              },
              child: Navigator(
                key: _nestedNavigatorKey,
                onGenerateRoute: (RouteSettings settings) {
                  return MaterialPageRoute<void>(
                    settings: settings,
                    builder: (_) => _buildNestedPage(0),
                  );
                },
              ),
            ),
          ),
        ],
      ),
    );
  }

  void _toggleCanLeave() {
    setState(() {
      _canLeave = !_canLeave;
      _status = _canLeave ? 'Route pop enabled' : 'Route pop disabled';
    });
  }

  void _pushNestedPage() {
    final int page = _nestedPage++;
    _nestedNavigatorKey.currentState?.push<void>(
      MaterialPageRoute<void>(
        settings: RouteSettings(name: 'nested-$page'),
        builder: (_) => _buildNestedPage(page),
      ),
    );
    setState(() {
      _status = 'Nested page $page pushed';
    });
  }

  static Widget _buildNestedPage(int page) {
    return Container(
      color: page == 0 ? const Color(0xFFE8F0FE) : const Color(0xFFE6F4EA),
      padding: const EdgeInsets.all(16),
      alignment: Alignment.center,
      child: Text(
        page == 0 ? 'Nested root' : 'Nested page $page',
        style: const TextStyle(fontSize: 18, color: Colors.black),
      ),
    );
  }
}
